### DISCUSSION TITLE
Fallback mod information from AssemblyInfo.cs if no BepInPlugin attribute is given

### DISCUSSION THREAD

**Discussion by Measurity** [BepInEx/BepInEx#209] [BepInEx 6 proposals]
**Is your feature request related to a problem? Please describe.**
I would like to reduce redundancy and tedium in project configuration.
The issue with using attributes is that I cannot change those fields at compile time without string replacing the code.

**Describe the solution you'd like**
Take the information required by the BepInEx "ChainLoader" class from the Assembly if it isn't found by the BepInPlugin attribute.
The GUID could also be taken from AssemblyInfo. Example definition: `[assembly: Guid("a4df9f47-b2d9-49a9-b237-09220857c051")]`.

**Describe alternatives you've considered**
 - Adding build step to string replace code files with the right BepInEx metadata.
 - Adding a batch file to rename the plugin metadata.

**Additional context**
It's possible that one solution has multiple BepInEx plugins defined so attribute should still be possible and have priority over AssemblyInfo.

**Improvements and extras**
Generate GUID if not defined by using author+mod name bytes into GUID constructor.

**Comment by js6pak** [comment]
#154

**Comment by ghorsington** [comment]
Greetings!

This is a good ticket in a sense that we have discussed this before but have never really written down our discussions on a public channel.  
It has been a while since we have talked about this, so I will try to summarise our rationale regarding this feature as best as I remember and can find from the BepInEx Discord server. I hope that this might also prompt some new discussions on the issue at hand since last talks lead to no changes.

To preface: all feature requests are always by default regarding upcoming new BepInEx versions. As such, this discussion will be related to BepInEx 6 development as BepInEx 5 is in Long Term Support mode and thus will not receive major changes to its API.

***

Just like with any plugin loader, plugin metadata is important as it allows plugin loaders to discern miscellaneous assemblies and classes from those that are of interest. On the other hand, metadata allows to subdivide code into reasonable chunks and tell the loader useful info about how to load the plugin. In case of BepInEx, plugin metadata is

* What classes should be considered as a loadable plugin
* Unique but readable plugin ID using [reverse domain notation](https://eqdn.tech/reverse-domain-notation/)
* Human-readable plugin name
* Plugin version as valid [semver 2.0](https://semver.org/) version
* Other plugins that the plugin depends on (as plugin ID + [npm semver range](https://docs.npmjs.com/cli/v6/using-npm/semver#ranges))
* Processes that the plugin should be loaded on (so-called process filters)

An important part of this is that this metadata exists for a plugin class. Just like you can have multiple classes in an assembly, you can thus also have multiple plugin classes in the same assembly. This has been used some cases alongside process filters so that you can load only specific plugins for each process but still ship them in the same DLL.

Moving metadata from classes to assembly has thus its own problems. Most notable of those are:

1. Different plugin classes require different metadata
    Oftentimes each `BepInPlugin` of a class is unique: each plugin has its own name, ID and version. In fact, each plugin class **must** have a unique ID. Thus if you have an assembly like

    ```cs
    // Which plugin does this GUID belong to???
    // In addtion, this is not a valid plugin ID!
    [assembly:Guid("a4df9f47-b2d9-49a9-b237-09220857c051")]

    [BepInPlugin(/*...*/)]
    class Plugin1 : BasePlugin { /* ... */ }
    
    [BepInPlugin(/*...*/)]
    class Plugin1Variation : BasePlugin { /* ... */ }
    ```
    Same applies for example to plugin dependencies and process filters.
   
2. What metadata is reasonable to move to assembly level?
    Related to the previous one, the question is: what metadata can and should be moved to assembly level?
    Is it just the human-readable name? Is it just the version? Should dependencies also be specified that way?

3. How the assembly-level data should be mapped to plugin-level?
    For example, normal .NET assembly versioning does not follow semver format which means that reading plugin version from default .NET attributes. 

    What if both plugin-level metadata and assembly-level meatadata is missing (e.g. you don't specify version in neither `BepInPlugin` or assembly attribute)? Right now `BepInPlugin` enforces you to specify the metadata which allows proper handling of it. That way there is a less chance of having plugin with version `0.0.0.0`.

    Does this need a custom attribute? If so, it adds API surface and has potential of confusion. Ability to do thing only one way in BepInEx allows for less confusion.

These are the ones I remember from our initial discussions. @bbepis and @ManlyMarco could add in their own opition on the topic since I might misremember something.

While there is #154, we haven't merged it as we still are still not in an agreement on this. As such, I hope for more discussion on this topic for solid solutions on raised issues.

**Reply by Therzok** [reply]
To be fair, I don't like the approach of having a GUID, inferring it from context (assembly/type) and csproj metadata (Version, AssemblyName, Company, etc)  is better

Harmony recently added reverse patch lookup.

Having no way to trace where patches originated from (every plugin is a GUID) is going to make this a headache and will require a diagnosis tool written just to understand what plugin is causing issues.

**Reply by ghorsington** [reply]
If by GUID you mean the normal .NET [Guid](https://docs.microsoft.com/en-us/dotnet/api/system.guid?view=net-5.0), I agree since its verbosity is not usable at all for plugins.

As for plugin IDs, they are useful since we don't want to assume assembly and type names to be static. Keeping IDs decoupled from C# syntax allows one to easily reference them and even move implementations when needed. Plugin IDs are unique and never changing for a plugin, no matter where it ends up attached to. A concrete example is someone making a plugin that also replaces features of another plugin (i.e. owing to deprecation of original one). This has already been used in a few mod loader plugins for BepInEx itself where plugins from another mod loader are given IDs so they can be referred to from other BepInEx plugins without caring where the plugin resides or if it has been shimmed or its type hierarchy changed because of shimming.

Not to mention how even right now plugin IDs are used with mod managers.

I'll trust general assembly metadata to display to user, but not class structure. So far all successful mod loader solutions I've seen provide separate ID for the mod instead of inferring it from refactorable class structure.

I can say for certain right now that decoupled plugin IDs will continue to be a thing and that they will continue being a constant. The only real question for debate is how to assign it to a plugin.

**Reply by Therzok** [reply]
Also, GUIDs will make it pretty complicated to work with dependencies.

**Comment by Measurity** [comment]
Preface: this answer differs from initial issue to incorporate issues brought up by ghorsington.

### Short answer
Have an API similar to Harmony where you specify *how* the plugins are discovered, through code instead of attributes. Either by requesting search of assembly X for all `BepInExPlugin` attributes (like now, but third-party controls it) or specifying plugins by type.

### Long answer
I propose an API to define how the assembly should be loaded which should be internally used by BepInEx to load plugin meta data from attributes as well. BepInEx would discover an "entrypoint" in my assembly and invoke it (similar to the [BepInExPlugin]). Then it executes the following pseudo code in my assembly that registers the plugins that I want:

```cs
BepInExLoader
    .WithPlugin(typeof(MyType), new PluginMetaData(NAME, VERSION, GUID))
    .WithPlugin(typeof(MyOtherType)) // Will use attribute or error out if not defined.
    .WithPlugin(typeof(MainPlugin), PluginMetaData.FromExecutingAssembly())
    .WithAssembly(Assembly.LoadFrom(...)) // Loads all BaseUnityPlugin/BasePlugin types for you, requires attributes on plugins to work. If assembly contains single plugin, allow fallback to AssemblyInfo.
    .Execute();
```

To get the same functionality as now I could use it like this:
```cs
BepInExLoader.WithAssembly(Assembly.GetExecutingAssembly()).Execute();
```

**Answering your points:**

> What classes should be considered as a loadable plugin

See pseudo code.

> Unique but readable plugin ID using reverse domain notation

I don't know why this is important. To me it seems redundant information because you have author + mod name. The purpose of an ID is to be unique, it doesn't have to be understandable. If this information is required I suggest adding a new field to specify.

> Human-readable plugin name

Could be provided by meta data or (as a fallback) inferred from class name.

> Plugin version as valid semver 2.0 version

Can be done in AssemblyInfo.cs as well.
See: <https://kimsereyblog.blogspot.com/2018/04/sementic-versioning-for-dotnet.html>
`[assembly: AssemblyInformationalVersion("2.0.0+24.Branch.master.Sha.6e4b39ab28bb63227d9c418a9460fe344c0bacb2")]`

> Other plugins that the plugin depends on (as plugin ID + npm semver range)

Dependants can use unique GUID (with comment of which mod it is). Imposing readability on GUID has maintainability cons as mention in this ticket. Optionally provide an NPM semver range which BepInEx would parse and validate while loading.

> Processes that the plugin should be loaded on (so-called process filters)

Does not conflict with current suggestion. Could integrate into pseudo code or leave as attribute.

> What if both plugin-level metadata and assembly-level meatadata is missing (e.g. you don't specify version in neither BepInPlugin or assembly attribute)? Right now BepInPlugin enforces you to specify the metadata which allows proper handling of it. That way there is a less chance of having plugin with version 0.0.0.0.

Could be enforced if necessary. Definitely should decide but it's an easily solved problem.

### Other questions that were asked elsewhere

> Wouldn't this break ILmerged mods?

Yes but it can be solved again if BepInEx allows multiple "entrypoints" defined in assemblies. The main idea is registering BepInEx plugins by code instead of letting attributes define this. BepInEx would invoke all "entrypoints" and the callee would register the plugins through the API.

**Reply by Therzok** [reply]
The C# way of defining custom `EntryPoint`s is via Attributes and reflection from assemblies. I'm not sure what this proposal here tries to solve.

How would BepInEx resolve the `main entrypoint`? It would still have to look for an Attribute that contains the `main entrypoint`.

The main problem is that:

* This opens up the implementation of the plugin chain loader to the user (assembly resolve, management of plugin types)
* Adds a lot more code to implement it
* Has little value, as this can be achieved as an extension library that has a Main plugin and does its own loading logic.

**Reply by Measurity** [reply]
> The C# way of defining custom EntryPoints is via Attributes and reflection from assemblies. I'm not sure what this proposal here tries to solve.

Make BepInEx nicer to use without jumping through hoops.

> How would BepInEx resolve the main entrypoint? It would still have to look for an Attribute that contains the main entrypoint.

Yes but it is not the point of the proposal. I would like more control over how the plugin info is retrieved without string replacing my .cs files at build time.

> This opens up the implementation of the plugin chain loader to the user (assembly resolve, management of plugin types)

This is a good thing. On the BepInEx side it can verify if everything is set proper after the call to the mod happens.

> Adds a lot more code to implement it

True for BepInEx. Not necessarily for mod authors. BepInEx can fallback to using plugin attributes as it is now if no "entrypoint" is defined.

> Has little value, as this can be achieved as an extension library that has a Main plugin and does its own loading logic.

True, but it complicates installation process for users if BepInEx does not support it. You're basically asking me to implement the plugin loading and dependency resolving, without conflicting with BepInEx.

**Comment by ghorsington** [comment]
Moved feature proposals to Discussions, I'll try to comment the latest comment soon.

**Comment by ManlyMarco** [comment]
In my opinion this proposal so far adds way too much complexity and edge cases to deal with.

Adding plugins through code will complicate any external tools like mod managers or updaters, since now you will have to parse the IL and stimulate execution or actually execute this code in some way, risking side-effects and other breakage.

It will also complicate dependency handling and will now require loading and executing each assembly to see if plugins are loaded or not (unless you still use attributes for dependency declarations, which is against the goal). This can take a fairly significant amount of time when the amount of plugins reaches 100s. Currently this is cached because the attributes are deterministic.

This approach is close to what preloader patchers are in BepInEx 5 - you can do whatever in the initialize method, including bootstrapping your plugins. We already discussed limiting patchers in some way to make them deterministic and allow caching of their results, because they can become a significant startup cost way faster than normal plugins.

Overall what I think should be prioritized in this case is performance and ease of use/integration. For performance the plugin loading step needs to be deterministic in order to allow for as much caching as possible, while for ease of use the plugin metadata has to be statically available so that plugin managers/updaters/etc. can do their job. Having to add extra attributes like `AssemblyInformationalVersion` or requiring an external tool in the build makes a plugin harder to write for beginners and adds more things that can go wrong.

> The issue with using attributes is that I cannot change those fields at compile time without string replacing the code.

I'm not sure what you need to modify? Dependency versions can be made to automatically update in the attributes, and you'll want to manually update the plugin version anyways. If you still need to edit the attributes then the most reliable way is to use cecil in a post compile step.

Sorry I don't give any suggestions on how to make this work, but I think we exhausted them during previous discussions.

**Reply by Measurity** [reply]
Interesting points and I agree it's getting too complex to integrate everything. Using Cecil as post build step would be suitable for me. Just for the sake of discussion gonna to add this:

Even though I'm not familiar with how the caching works, a possible solution to cache plugins would be to do it once by code and then, by default, cache the information given by the mod unless mod author disables it explicitly. In which case it would run every time on startup.

Thanks guys.

**Comment by ghorsington** [comment]
One possible way to consider is not a cecil build step but rather a custom MSBuild target.

Right now I know of this:
https://github.com/Therzok/dsp_modding/blob/main/msbuild/BepInEx.targets 

Which reads assembly info off csproj and then exposes it as a `ThisAssembly` class that can be then used in plugin meta:

https://github.com/Therzok/dsp_modding/blob/main/src/DumpAssets/DumpAssetsPlugin.cs#L22

I'm personally fine with this approach to fulfil the original request for the ability to grab plugin meta from assembly data. I think approaching this via a MSBuild target is better because

* BepInEx core principles aren't changed, normal users and newcomers can use the normal direct attribute syntax to get started fast
* The issue of multiple plugins per assembly is delegated to the developer since he is given the metadata from csproj
* It's optional: MSBuild target can be a NuGet package, for instance

I think this would satisfy all needs presented here. Instead of having this be in BepInEx core, it can be a separate project. We could even provide ready plugin templates that use the custom build target.

I'm not sure if we should roll our own build target based on Therzok/dsp_modding. Is there anyone who would be willing to maintain it?

**Reply by Therzok** [reply]
@ghorsington my long term plan was releasing a .NET SDK that would have all the targets defined implicitly, but that's gonna take some time to actually finish.

**Comment by js6pak** [comment]
I feel like Therzok/dsp_modding's way of adding a file is really dirty. Mono.Cecil post build step feels much nicer (ignoring how annoying it is to bundle msbuild targets with nuget, but that applies to both approaches): https://github.com/NuclearPowered/Reactor.BepInAutoPlugin

**Reply by Therzok** [reply]
It's not annoying to bundle them with NuGet, you just need a `build` directory that has `NuGetPackage.targets`.

**Reply by Therzok** [reply]
> I feel like Therzok/dsp_modding's way of adding a file is really dirty.

The file is generated in the intermediate output path. Actually, this is how AssemblyInfo.cs is generated in C#. If BepInEx supported assembly attributing, it could happen without having the `ThisAssembly` file generated.

If we could have some annotations that work without affecting user code, that would be great. But there's no way to _automatically_ generate something that is going to be applied to `BepInExPluginAttribute`, except if we make it take its default values from the well known constants and use reflection.

On the other hand, for the Version, it's pretty easy for BepInEx to reflect `AssemblyVersion` or `AssemblyFileVersion` and `AssemblyName` attributes.

**Reply by Therzok** [reply]
The reason for codegen in `dsp_modding` is mostly because I have more properties available there, that I'm planning to use one way or another.

**Reply by Therzok** [reply]
But yeah, a generated file with just `[assembly: MyAttribute]` is something that doesn't touch the repo in any way.

What is dirty about generating a file in this scenario specifically?

**Reply by Therzok** [reply]
Think:
```csharp
public class AssemblyBepInPluginAttribute
{
    public AssemblyBepInPluginAttribute(Type type, string guid, string name, string version)
    {}
}
```

and the code is generated exactly like `AssemblyInfo.cs` with `[assembly: AssemblyBepInPlugin(...)]`

**Reply by ghorsington** [reply]
Hmmm, I don't think this attribute was discussed yet. It might warrant some thought.

**Comment by ghorsington** [comment]
I'll bump this into a new thread so that it won't get drowned:

One proposal was to add an assembly level attribute akin to this:

```cs
public class AssemblyBepInPluginAttribute
{
    public AssemblyBepInPluginAttribute(Type type, string guid, string name, string version)
    {}
}
```
(Courtesy of @Therzok)

That would then contain the type + relevant info. This would allow code generators to emit the assembly attribute fairly easily and this still would allow to have static data.

**Reply by Therzok** [reply]
`AttributeTargets.Assembly` should be specified.

**Reply by js6pak** [reply]
What would happen to `BepInPluginAttribute` in this case? Would it be totally omitted?

**Reply by js6pak** [reply]
This would also require https://github.com/dotnet/msbuild/pull/6285 which will drop with .net 6 sdk
